### ISSUE TITLE
Themes which use font-weight shift cursor position

### ISSUE THREAD

**Issue by smingerson** [gadenbuie/rsthemes#80]
Themes which increase the font-weight of certain elements, such as a11y-dark (and light, I assume), lead to cursor shifting in lines which contain such elements. This is maybe more of a problem with RStudio than rsthemes, but having these themes be fixed would be excellent. As far as I can tell, the font used and zoom level make no difference.

I 'fixed' the problem by searching for all instances of font-weight and commenting them out. 

For example:
1. Set theme to a11y-dark. Other themes which use font-weighting may exhibit problems in different places.
2. Create the following rmarkdown, or open another and examine a similar areas. 
````r 
---
title: "This bug makes me sad"
output: 
  html_document
---
# A heading 
A heading
Notice how the cursor is in slightly different positions relative to the "g" in heading when at the end of each line. 
However, as headings get longer, the problem gets worst. So, if you can't tell the difference, make a longer heading

Next, a chunk.

```{r my-r-chunk, echo = FALSE, eval = FALSE}
# Notice it also happens with function calls. The more function calls on a line, the greater the discrepancy.
it(also(happens(with(fnction(calls())))))
```

````

RStudio information
RStudio 2021.09.1+372 "Ghost Orchid" Release (8b9ced188245155642d024aa3630363df611088a, 2021-11-08) for Windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.8 Chrome/69.0.3497.128 Safari/537.36

**Comment by gadenbuie** [comment]
I've seen something similar before (so I know what you mean) but I can't reprex this particular issue. I'm using the same version of RStudio on Mac and get these results:

Start of line: ![image](https://user-images.githubusercontent.com/5420529/144909426-41832bd5-64e7-4139-b1d4-db69f7d10dd2.png) 

End of line: ![image](https://user-images.githubusercontent.com/5420529/144909487-9de433a8-3390-4879-9e57-bae124161fb1.png)

Function calls:  ![image](https://user-images.githubusercontent.com/5420529/144909644-a42c3fcf-f6d6-482a-afef-84d35b70a789.png)

What font are you using? I think the problem may be that you're using a font with a bold face that's wider than the normal face. I tried a few fonts to see if I could reprex the issue that way, but they all behaved as expected (Courier, Andale Mono, Fira Code, Hasklig, Monaco). Or maybe it's a Windows thing?

**Comment by smingerson** [comment]
Ah, I hadn't considered it being Windows-specific, but it would not surprise me. It happens with Fira Code on Windows, along with Courier, Courier New and Cascadia Mono.  

Here's some screenshots, all with the cursor positioned at end of line.
Courier font
![image](https://user-images.githubusercontent.com/31328867/144914048-74ce6bed-cb02-4257-9fec-a713cdc0bd4d.png)
![image](https://user-images.githubusercontent.com/31328867/144914112-87686bb3-7e7a-44b0-af47-0621c92027fa.png)
![image](https://user-images.githubusercontent.com/31328867/144914154-2b37175d-064b-4f60-8291-c050a5756ba9.png)

**Comment by gadenbuie** [comment]
Yeah the bold font glyphs are clearly wider than their normal font weight counterparts in those lines. Can you try something for me just to see what happens? Under Global Options > General > Advanced > OS Integration ... if you have similar options on Windows does the behavior change with any of them (and what's your current setting)?

![image](https://user-images.githubusercontent.com/5420529/144915654-559616b1-d486-41aa-8aa1-26b1ec825f08.png)

**Comment by Darxor** [comment]
I've also had this issue and tried to investigate it through - it seems there is no universal fix for this, because it lies deeply in how ACE editor (which RStudio uses) works. It calculates fontmetrics once, and uses them everywhere. So it can only work with only one **monospaced** font.
You can try to reproduce this issue on live demo at ACE website https://ace.c9.io, by changing e.g. fontfamily for comments or function calls.

However, there is a workaround - **you have to use a font, that has a native bold version installed on the system**. Say you use Fira Code, it comes with Regular, Light, Medium, Bold, SemiBold and Retina versions.
If you use "Regular" version (labelled Fira Code in RStudio), there will be a Bold variant associated with it, and it will work fine - since authors of Fira Code made sure different versions have the same character width. But, if you use "Retina" or "SemiBold" - there is now no "Bold" variant for those, so WebKit simulates them for you. This leads to different character widths between your base and bold variants.

TL;DR Install a font with native Bold version, that has the same character width as Regular, and use the Regular version.

**Comment by gadenbuie** [comment]
Thanks for the advice and thorough explanation @Darxor!

**Comment by smingerson** [comment]
I had 2 versions each of Fira Code font face installed. I uninstalled them all, and re-installed the most recent version -- now it's working as expected. Thanks for the explanation @Darxor, which led me to take another look!
